ingest/clusters.py

```diff
@@ -1,45 +1,77 @@
-import copy
-from typing import Dict
+from typing import Dict, Generator, List, Tuple, Union  # noqa: F401
+from dataclasses import dataclass
+from mypy_extensions import TypedDict
 
-from ingest_files import IngestFiles
+from ingest_files import DataArray
+from annotations import Annotations
 
 
-class Clusters(IngestFiles):
+@dataclass
+class DomainRanges(TypedDict):
+    x: List
+    y: List
+    z: List = None
 
-    ALLOWED_FILE_TYPES = ["text/csv", "text/plain", "text/tab-separated-values"]
-    COLLECTION_NAME = "clusters"
+
+class Clusters(Annotations):
+    # Need to delete after refactor
+    COLLECTION_NAME = "cluster_groups"
     SUBCOLLECTION_NAME = "data"
 
+    ALLOWED_FILE_TYPES = ["text/csv", "text/plain", "text/tab-separated-values"]
+    LINEAR_DATA_TYPE = 'ClusterGroup'
+
+    @dataclass
+    class Document(TypedDict):
+        name: str
+        # 3d or 2d cluster
+        cluster_type: str
+        # Array of Hash objects that describe all extra "annotation" columns
+        cell_annotations: List
+        file_id: str
+        study_id: str
+        # Hash containing min/max arrays for each axis in the cluster plot
+        domain_ranges: DomainRanges = None
+
     def __init__(
         self,
         file_path: str,
-        file_id: str,
-        study_accession: str,
-        *,
+        study_id: str,
+        study_file_id: str,
         name: str,
+        *,
         domain_ranges: Dict = None,
     ):
-        IngestFiles.__init__(self, file_path, self.ALLOWED_FILE_TYPES)
-        self.header = self.get_next_line(increase_line_count=False)
-        # Second line in cluster is metadata_type
-        self.metadata_types = self.get_next_line(increase_line_count=False)
+        Annotations.__init__(self, file_path, self.ALLOWED_FILE_TYPES)
+        self.determine_coordinates_and_cell_names()
         self.source_file_type = "cluster"
-        self.has_z = True if ("z" in self.header or "Z" in self.header) else False
-        self.cell_annotations = self.create_cell_annotations_field()
+        self.cluster_type = (
+            '2d'
+            if (
+                "z" in self.coordinates_and_cell_headers
+                or "Z" in self.coordinates_and_cell_headers
+            )
+            else '3d'
+        )
+        # self.cell_annotations = self.create_cell_annotations_field()
         self.cluster_subdocs = {}
+        self.name = name
+        self.domain_ranges = domain_ranges
+        self.study_id = study_id
+        self.study_file_id = study_file_id
         # TODO: Populate the cell_annotations array when pandas is implemented
         self.top_level_doc = {
             "name": name,
-            "cluster_type": "3d" if self.has_z else "2d",
             "cell_annotations": [],
-            "study_accession": study_accession,
+            "study_accession": study_id,
```

Comment:
study_accession and study_id are not the same thing.  The accession is SCP123, whereas the study_id is a MongoDB document id (like 5d276a50421aa9117c982845).

Reply:
This is irrelevant right. The code is still here untill I make the final transition.